### PR TITLE
fuzz: implement run_afl, add seeds for fuzzer

### LINKED ISSUE #3905: Support AFL fuzzer
## Feature Request

In PR https://github.com/tikv/tikv/pull/3848, we introduced multi-fuzzer framework. However for AFL fuzzer, we additionally need initial seek to make it work.

### PR DESCRIPTION
**Pull request by aylei**
~This issue is WIP. I open this PR so to discuss how should we generate seed for AFL.~

I generate 64 random bytes manually as the seed now, I want to know if it is necessary to generate seeds on the fly or take a step forward, allowing users to specify seeds in the test code on their own.  

support AFL fuzzer, fix #3905

Signed-off-by: alei <rayingecho@gmail.com>

## What have you changed? 
Support AFL fuzzer, add testcase (as seed) for AFL fuzzer.

## What are the type of the changes? 
- New feature

## How has this PR been tested? 
Tested manually.

## Does this PR affect documentation (docs) update?
Yes (but currently haven't)

## Does this PR affect tidb-ansible update?
No

### CODE REVIEW

**Comment by breeswish** [comment]
Thanks for your contribution! I think a good choice is to use some manually generated seeds. This can be very effective. How do you think @overvenus

**Comment by overvenus** [comment]
> to use some manually written seeds (i.e. we maintain a directory containing these seeds for different fuzz targets). 

It's good idea, and I think it's a future work.

**Comment by aylei** [comment]
> to use some manually written seeds (i.e. we maintain a directory containing these seeds for different fuzz targets).
Agreed. I will add manually written seeds support, and if a target has no seeds provided, use the default `in` directory.
As a test, providing the default random seed, all the AFL fuzz targets can complete the first cycle within 2 minutes, which is fairly enough currently.

BTW, AFL fuzzing procedure has to be terminated explicitly (otherwise, AFL will run permanently). I don't know if I should add a timeout for fuzzing.

**Comment by breeswish** [comment]
@aylei Yeah, please keep it running. Our other fuzzers keep running as well! 😄

File: fuzz/README.md [hunk 1]
```diff
@@ -36,7 +33,11 @@ See [honggfuzz-rs documentation](https://github.com/rust-fuzz/honggfuzz-rs).
 cargo install afl
 ```
 
-See [the fuzz.rs book](https://fuzz.rs/book/afl/setup.html).
+Seed files should be placed in the `fuzz/fuzzer-afl/seeds/{target}/` directory, where `target` is the fuzz target name.
+
+If no seed file provided, `fuzz/fuzzer-afl/seeds/default/` will be used as seeds.
```
Comment (breeswish) [inline fuzz/README.md]: Maybe we can also share seeds for different fuzzers. You can have a try (not mandatory) :)

**Comment by breeswish** [comment]
This PR looks promising. @overvenus Please take a look as well.

File: fuzz/cli.rs [hunk 2]
```diff
@@ -161,12 +161,58 @@ fn run(fuzzer: Fuzzer, target: &str) -> Result<(), Error> {
...
-    // AFL requires initial inputs, so leave it to future.
-    // General process:
+fn run_afl(target: &str) -> Result<(), Error> {
+    let fuzzer = Fuzzer::Afl;
+
     // 1. cargo afl build (in fuzzer-afl directory)
-    // 2. cargo afl fuzz -i in -o out target/debug/fuzzer-afl
-    unimplemented!()
+    let fuzzer_build = Command::new("cargo")
+        .args(&["afl", "build", "--bin", target, "--target-dir", "./target"])
```
Comment (breeswish) [inline fuzz/cli.rs]: looks like `--target-dir` is not very necessary?

File: fuzz/README.md (hunk 1 above)
Comment (breeswish) [reply fuzz/README.md]: Examples:

https://github.com/rust-fuzz/targets/blob/master/cli.rs#L330
https://github.com/rust-fuzz/targets/blob/master/cli.rs#L226

Comment (aylei) [reply fuzz/README.md]: looks good, I will give a try

File: fuzz/cli.rs (hunk 2 above)
Comment (aylei) [reply fuzz/cli.rs]: without specifying target dir, this crate will share the `target` dir of the top level workspace, which is not expected IMHO.

Comment (breeswish) [reply fuzz/cli.rs]: Yeah, in fact I was expecting that, but I don't have a strong opinion, as long as gitignore files are handled correctly.

**Comment by aylei** [comment]
@breeswish amend based on review comments

File: fuzz/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+fuzzer-*/corpus-*
```
Comment (breeswish) [inline fuzz/.gitignore]: Prefer to keep a new line.

File: fuzz/cli.rs [hunk 4]
```diff
@@ -160,13 +161,74 @@ fn run(fuzzer: Fuzzer, target: &str) -> Result<(), Error> {
...
+    let seeds_dir = SEED_ROOT.join(target);
+    if seeds_dir.exists() {
+        seeds_dir
+    } else {
+        SEED_ROOT.join("default")
+    }
+}
+
+/// Create corpus dir for fuzz target
+fn create_corpus_dir(base: &PathBuf, target: &str) -> Result<PathBuf, Error> {
```
Comment (breeswish) [inline fuzz/cli.rs]: How about `base: impl AsRef<Path>`? Then your calling to `create_corpus_dir` can be simplified.

File: fuzz/cli.rs [hunk 5]
```diff
@@ -176,8 +238,14 @@ fn run_honggfuzz(target: &str) -> Result<(), Error> {
     let mut rust_flags = env::var("RUSTFLAGS").unwrap_or_default();
     rust_flags.push_str("-Z sanitizer=address");
 
-    let mut hfuzz_args = env::var("HFUZZ_RUN_ARGS").unwrap_or_default();
-    hfuzz_args.push_str("--exit_upon_crash --run_time 5");
+    let hfuzz_args = format!(
+        "-f {} \
+         --exit_upon_crash \
+         --run_time 5 \
```
Comment (breeswish) [inline fuzz/cli.rs]: This is a mistake introduced by me. Looks like "run_time 5" is not needed at all. You can help change it (or leave to me XD)

File: fuzz/cli.rs [hunk 6]
```diff
@@ -232,6 +302,8 @@ fn run_libfuzzer(target: &str) -> Result<(), Error> {
 
     let fuzzer_bin = Command::new("cargo")
         .args(&["run", "--target", &target_platform, "--bin", target])
+        .arg(&corpus_dir)
```
Comment (breeswish) [inline fuzz/cli.rs]: I guess we need "--" args, so that it will be:

```bash
cargo run --target target --bin bin -- corpus_dir seed_dir
```

**Comment by aylei** [comment]
@breeswish Thanks for your patient review!

**Comment by breeswish** [comment]
@overvenus PTAL

**Pull request merged**